### DISCUSSION TITLE
Lesson 7 Challenge: cast send works, but script throws error

### DISCUSSION THREAD

**Discussion by RichuAK** [Cyfrin/foundry-full-course-cu#689] [Q&A] [answered]
Hello, I tried to solve the challenge in Lesson Seven, but the script I'm running is throwing an error.

Here's the script I'm trying to run:

```solidity
function run() external {
        uint256 privateData = getPrivateData();
        console.log("The Data: ", privateData);
        console.log("Attempting to find Solution for Lesson Seven");
        console.log("From address: ", msg.sender);

        LessonSeven lessonSeven = LessonSeven(LessonSevenAddress);
        lessonSeven.solveChallenge(privateData, "richuak1");
    }
```

I'm omitting the definition of ```getPrivateData()``` in the spirit of the challenge, but the function returns the correct value.

I am able to solve the challenge by using ```cast send``` with the data, but when I'm trying to run the above script, it throws an error called  ```ERC721: transfer to non ERC721Receiver implementer``` 

Here's the full error:

<img width="1115" alt="Screenshot 2023-08-22 at 14 31 57" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17203781/88762507-3db1-4f0a-a568-337e34c6c89a">

Would love it if I can understand why it's throwing the error. The account I'm sending from is an EOA, not a contract.

Thanks in advance!

**Comment by Niferu** [comment, marked as answer]
Hello @RichuAK 

Logs are currently telling you why this transfer is failing :) `ERC721: transfer to non ERC721Receiver` that's all you need, keep digging, good luck

**Comment by RichuAK** [comment]
Hey @Neftyr, 

Thanks for the tip!

I think it was because while I was thinking the transaction was being sent by the EOA (thereby eliminating any ```_safeMint()``` error possibilities), the actual ```solveChallenge()``` method was being hit by my own script contract, which doesn't have an ```onERC721Received``` function implemented for ```safeMint```.

So I ran the script again with ```prank``` :

```solidity
function run() external {
        uint256 privateData = getPrivateData();
        console.log("The Data: ", privateData);
        console.log("Attempting to find Solution for Lesson Seven");
        console.log("From address: ", msg.sender);
        vm.prank(msg.sender);
        LessonSeven lessonSeven = LessonSeven(LessonSevenAddress);
        lessonSeven.solveChallenge(privateData, "richuak1");
    }
```

And this time it ran without any errors:
<img width="1113" alt="Screenshot 2023-08-22 at 20 49 28" src="https://github.com/Cyfrin/foundry-full-course-f23/assets/17203781/7aa52ce1-9755-4c8a-8448-4c001e6ff5e0">

But the transaction is not showing up on Etherscan, which was a surprise for me. It was only then I realised that it's a script and not a deployed contract, so the whole thing is a virtual transaction? The logs and the event emitting, all virtual.

This is my current understanding of why the transaction isn't showing up on Etherscan. Let me know if I'm wrong in this.

Again, thanks for the tip!

**Reply by Niferu** [reply]
@RichuAK 

Good job bro, if you are using prank transaction will not hit blockchain, in order to see tx on blockchain you should use `vm.startBroadcast();`, see below :)

https://book.getfoundry.sh/cheatcodes/broadcast?highlight=start%20broadcas#broadcast

**Reply by RichuAK** [reply]
@Neftyr 

Beautiful! Transaction hits the blockchain and everything works perfectly! 

Thanks a ton bro!

**Reply by Niferu** [reply]
@RichuAK 

I'm very glad I could help, no problem Sir and congratz again you solved challenge yourself! :)
